### PR TITLE
raft: support IP address change

### PR DESCRIPTION
**Pull request by kostja**
This is the core of dynamic IP address support in Raft, moving out the IP address sourcing from Raft Group 0 configuration to gossip. At start of Raft, the raft id <> IP address translation map is tuned into the gossiper notifications and learns IP addresses of Raft hosts from them.
The series intentionally doesn't contain the part which speeds up the initial cluster assembly by persisting the translation cache and using more sources besides gossip (discovery, RPC) to show correctness of the approach.

### CODE REVIEW

File: service/raft/raft_group_registry.cc [hunk 1]
```diff
@@ -86,7 +164,8 @@ void raft_group_registry::init_rpc_verbs() {
             auto& rpc = self.get_rpc(gid);
             // The address learnt from a probably unknown server should
             // eventually expire
-            self._srv_address_mappings.set(from, std::move(addr), true);
+            rslog.debug("Adding address {} -> {} to the address map", from, addr);
```
Comment (kbr-) [inline service/raft/raft_group_registry.cc:167]: I was using the DEBUG level for `rslog` to debug stuff in the past. But this will make the DEBUG level too crowded for `rslog` (this is printed on **every** received message) and I'll never want to enable it again. I think this should be `trace` at most, or even not happen at all. Is it that useful?

**Review by kbr- — CHANGES_REQUESTED** [review]

File: service/raft/raft_address_map.hh [hunk 2]
```diff
@@ -263,101 +262,86 @@ public:
...
-        return entry._addr;
+        add_expiring_entry(it->first, entry);
+        replicate_set_expiring(it->first);
     }
+    // Inserts a new expiring mapping or updates the existing one
+    // with the new IP address. Never clears the expiring property
+    // from an existing entry.
+    // @param update_if_exists update the entry if it exists
+    // @return true if the entry is expiring
+    bool add_or_update_entry(raft::server_id id, gms::inet_address addr, bool update_if_exists = true) {
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_address_map.hh]: I removed the return value already in a follow up. update_if_exists is moved in the next patch, raft-ip-final, when we prompt the address map with discovery and other sources.

File: service/raft/raft_group_registry.cc [hunk 3]
```diff
@@ -322,6 +414,19 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
     return _direct_fd_proxy;
 }
 
+void
+raft_group_registry::notify_failure_detector(raft::server_id id, bool track) {
+    if (!_direct_fd_subscription.has_value()) {
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_group_registry.cc:419]: I think it's a side effect of moving things around a few times.

File: service/raft/raft_address_map.hh [hunk 4]
```diff
@@ -172,50 +170,51 @@ class raft_address_map : public peering_sharded_service<raft_address_map<Clock>>
...
+    void replicate_set_expiring(const raft::server_id& id) {
+        replicate([id] (raft_address_map_t& self) {
+            self.handle_set_expiring(id);
         });
     }
 
-    void handle_set_nonexpiring_entry(const raft::server_id& id, const gms::inet_address& addr) {
-        auto [it, _] = _map.try_emplace(id, addr);
+    void handle_set_nonexpiring(const raft::server_id& id) {
+        auto [it, _] = _map.try_emplace(id, timestamped_entry{{}});
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_address_map.hh:216]: For what it's worth I don't believe the generated code will be any different. The compiler is allowed to optimize away this constructor + move and construct in place. But I will change it.

[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_address_map.hh:216]: The reason I use the type name here is that it adds clarity - we need to go back to the definition of the map and see what its type and hhow it's implicitly constructed from std::nullopt.

File: service/raft/raft_address_map.hh [hunk 5]
```diff
@@ -263,101 +262,86 @@ public:
     // - we need to remove id of the node self is replacing.
     std::optional<raft::server_id> find_replace_id(gms::inet_address addr, raft::server_id self) const {
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_address_map.hh:285]: Yes.

File: service/raft/raft_address_map.hh [hunk 6]
```diff
@@ -263,101 +262,86 @@ public:
...
         for (auto& [id, entry] : _map) {
-            if (entry._addr == addr && id != self) {
+            if (entry._addr.has_value() && entry._addr.value() == addr && id != self) {
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_address_map.hh:287]: I don't want to. std::optional has an overloaded operator== which accepts e.g. other optionals, optionals of other types, and an optional is implicitly-constructible. I find this utterly broken. I don't want this code to quietly compile whenever either type changes.

File: service/raft/raft_group0.cc [hunk 7]
```diff
@@ -86,6 +87,41 @@ namespace service {
...
+    explicit group0_rpc(raft_group_registry& raft_gr,
+            raft_state_machine& sm, netw::messaging_service& ms,
+            raft_address_map& address_map, raft::group_id gid, raft::server_id srv_id)
+        : raft_rpc(sm, ms, address_map, gid, srv_id)
+        , _raft_gr(raft_gr)
+    {}
+};
+
+void
+group0_rpc::on_configuration_change(raft::server_address_set add, raft::server_address_set del) {
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_group0.cc:107]: I don't want to give the compiler neither choice nor responsibility about deciding whether to inline this function. It's a virtual function, it can't be inline anyway - but generally  I avoid inline definitions for anything beyond trivial. Besides, today you inline it, but tomorrow it becomes too big to follow inline - or the code is moved out  to a header file, so the function has to be deinlined, which would create a diff noise. There is value in lower indent as well. 
Generally I don't think comments like that are worth making, well, unless you'd like to get to know other people's coding habits, or you're raising a point about an implicit yet fairly widespread scylla coding convention I am not aware of.

[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_group0.cc:107]: Honestly, I didn't even think when deciding whether to inline this code or not. I forget when I started to avoid inline definitions  - but it was long, long time ago.

[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_group0.cc:107]: I find vim  fold markers are extremely convenient for files which contain definitions for multiple classes. Fold markers are not equivalent to syntax folding, so moving the definition inline will not achieve the desired effect.

File: service/raft/raft_group0.cc [hunk 8]
```diff
@@ -348,7 +372,19 @@ future<> raft_group0::start_server_for_group0(raft::group_id group0_id) {
...
-
+    // At this time the group registry is already up and running,
+    // so the address map is getting all the notifications from
+    // the gossiper. By reading the persistent replica of
+    // application state *after* subscribing to new gossip events,
+    // we ensure we haven't missed any IP update in the map.
+    auto ids = co_await db::system_keyspace::load_raft_server_ids();
+    for (const auto& [ip_addr, id] : ids) {
+        auto server_id = raft::server_id{id};
+        _raft_gr.address_map().add_or_update_entry(server_id, ip_addr);
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_group0.cc:383]: Apparently this is done in the last patch and not cherry-picked, I will cherry-pick.

File: service/raft/raft_group_registry.cc [hunk 9]
```diff
@@ -65,11 +66,88 @@ class raft_group_registry::direct_fd_proxy : public raft::failure_detector, publ
...
+        const gms::versioned_value& newvalue) override;
+    virtual future<> on_change(gms::inet_address endpoint, gms::application_state state,
+        const gms::versioned_value& value) override;
+    virtual future<> on_alive(gms::inet_address endpoint, gms::endpoint_state ep_state) override;
+    virtual future<> on_dead(gms::inet_address endpoint, gms::endpoint_state state) override;
+    virtual future<> on_remove(gms::inet_address endpoint) override;
+    virtual future<> on_restart(gms::inet_address endpoint, gms::endpoint_state ep_state) override;
+};
+
+future<> gossiper_state_change_subscriber_proxy::on_endpoint_change(
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_group_registry.cc:93]: See above.

File: service/raft/raft_group_registry.cc (hunk 1 above)
Comment (kostja) [reply service/raft/raft_group_registry.cc:167]: I can drop it. If you can enable debug level log you can add extra logging to the source code.

File: service/raft/raft_group0.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (kbr-) [reply service/raft/raft_group0.cc:383]: I still think we should get rid of `update_if_exists` altogether to simplify the interface, and if you need a conditional update, use `find`. If I understand correctly, we only need conditional updates outside hot paths.

File: service/raft/raft_group0.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (kbr-) [reply service/raft/raft_group0.cc:107]: > I don't want to give the compiler neither choice nor responsibility about deciding whether to inline this function. It's a virtual function, it can't be inline anyway - but generally I avoid inline definitions for anything beyond trivial. 

Right now I'm looking at this purely from how-the-code-looks perspective.

I agree we shouldn't do inline definitions **when** there's a separate .hh and .cc file containing parts of the class.

But in this code, when the whole class sits in a single file, these out-of-line definitions is just repeating ourselves and making the code longer and harder to maintain. The only reason we must do this in general is the .hh vs .cc separation. Rust, for example, doesn't have out-of-line member function definitions at all. We'll also not need them once we migrate away from header files to C++ modules.

Anyway, it's not that important, I don't like the code repetition and the code length, but since it's a purely subjective thing, it's fine if it stays that way.

File: test/raft/raft_address_map_test.cc [hunk 10]
```diff
@@ -52,79 +52,60 @@ SEASTAR_THREAD_TEST_CASE(test_raft_address_map_operations) {
...
-        sharded<raft_address_map<manual_clock>> m_svc;
+        // To mark an entry non-expiring, call set_nonexpiring().
+        sharded<raft_address_map_t<manual_clock>> m_svc;
         m_svc.start().get();
         auto stop_map = defer([&m_svc] { m_svc.stop().get(); });
         auto& m = m_svc.local();
 
-        m.set(id1, addr1, false);
+        m.add_or_update_entry(id1, addr1);
+        m.set_nonexpiring(id1);
```
Comment (kbr-) [inline test/raft/raft_address_map_test.cc:94]: This is redundant, it's exactly the same as the first test now.

File: test/raft/raft_address_map_test.cc [hunk 11]
```diff
@@ -135,40 +116,31 @@ SEASTAR_THREAD_TEST_CASE(test_raft_address_map_operations) {
...
+        sharded<raft_address_map_t<manual_clock>> m_svc;
         m_svc.start().get();
         auto stop_map = defer([&m_svc] { m_svc.stop().get(); });
         auto& m = m_svc.local();
 
-        m.set(id1, addr1, false);
-        BOOST_CHECK_THROW(m.set(id1, addr2, false), std::runtime_error);
+        m.add_or_update_entry(id1, addr1);
+        m.set_nonexpiring(id1);
+        BOOST_CHECK_NO_THROW(m.add_or_update_entry(id1, addr2));
```
Comment (kbr-) [inline test/raft/raft_address_map_test.cc:128]: You should check that the update took effect (`!!m.find(id1) && *m.find(id1) == addr2)`). Right now you don't have a single test for whether updating existing entries changes them.

File: test/raft/raft_address_map_test.cc [hunk 12]
```diff
@@ -135,40 +116,31 @@ SEASTAR_THREAD_TEST_CASE(test_raft_address_map_operations) {
...
         auto stop_map = defer([&m_svc] { m_svc.stop().get(); });
         auto& m = m_svc.local();
 
-        m.set(id1, addr1, false);
-        m.set(id1, addr1, true);
+        m.add_or_update_entry(id1, addr1);
+        m.set_nonexpiring(id1);
+        m.add_or_update_entry(id1, addr2);
         manual_clock::advance(expiration_time);
         BOOST_CHECK(m.find(id1));
```
Comment (kbr-) [inline test/raft/raft_address_map_test.cc:142]: Could also check if the value is correct.

File: test/raft/raft_address_map_test.cc (hunk 11 above)
Comment (kbr-) [reply test/raft/raft_address_map_test.cc:128]: Also, I think you can get rid of the `BOOST_CHECK_NO_THROW`, the test would fail anyway if the call threw.

File: test/raft/raft_address_map_test.cc [hunk 13]
```diff
@@ -188,24 +160,26 @@ SEASTAR_THREAD_TEST_CASE(test_raft_address_map_replication) {
     using seastar::manual_clock;
 
     {
-        sharded<raft_address_map<manual_clock>> m_svc;
+        sharded<raft_address_map_t<manual_clock>> m_svc;
         m_svc.start().get();
         auto stop_map = defer([&m_svc] { m_svc.stop().get(); });
         auto& m = m_svc.local();
 
         // Replicate non-expiring entry, ensure it doesn't expire on the other shard
```
Comment (kbr-) [inline test/raft/raft_address_map_test.cc:168]: s/entry/flag; we don't replicate entries now, we only replicate the (non)expiration flag.
I'd reword to something like:
```
// Add entry on both shards, replicate non-expiration flag, ensure it doesn't expire on the other shard
```

File: test/raft/raft_address_map_test.cc [hunk 14]
```diff
@@ -214,21 +188,22 @@ SEASTAR_THREAD_TEST_CASE(test_raft_address_map_replication) {
         BOOST_CHECK(!m.find(id1));
 
         // Expiring entries are not replicated
```
Comment (kbr-) [inline test/raft/raft_address_map_test.cc]: This test has no meaning now.
Either entries are not replicated: expiring and non-expiring ones.
Previously it had meaning because some entries were replicated, some weren't; so we checked that those that weren't, indeed weren't.

You could update this comment (remove "expiring") or just delete the test altogether.

File: test/raft/raft_address_map_test.cc [hunk 15]
```diff
@@ -214,21 +188,22 @@ SEASTAR_THREAD_TEST_CASE(test_raft_address_map_replication) {
...
-        m.set(id1, addr1, true);
+        m.add_or_update_entry(id1, addr1);
         ping_shards().get();
-        m_svc.invoke_on(1, [] (raft_address_map<manual_clock>& m) {
+        m_svc.invoke_on(1, [] (raft_address_map_t<manual_clock>& m) {
             BOOST_CHECK(!m.find(id1));
         }).get();
 
         // Can't add non-expiring entries on shard other than 0
```
Comment (kbr-) [inline test/raft/raft_address_map_test.cc:197]: s/add non-expiring entries/set non-expiring flag/

File: service/raft/raft_group0.cc [hunk 16]
```diff
@@ -894,7 +894,7 @@ future<group0_peer_exchange> raft_group0::peer_exchange(discovery::peer_list pee
             co_return group0_peer_exchange{group0_info{
                 .group0_id = std::get<raft::group_id>(_group0),
                 .id = _raft_gr.group0().id(),
-                .ip_addr = _raft_gr.address_map().get_inet_address(_raft_gr.group0().id())
+                .ip_addr = _raft_gr.address_map().find(_raft_gr.group0().id()).value_or(gms::inet_address{}),
```
Comment (kbr-) [inline service/raft/raft_group0.cc:897]: Continuing our discussion from earlier https://github.com/scylladb/scylladb/pull/11635/commits/bf635b9d7e063d0b5c3aa32539b2b38a6dd643d5#r1019092762

> So do you suggest using get_broadcast_address instead?

Yes, actually, that would make my question irrelevant. Whether we have our own IP in the address map or not, we will always be able to return our own IP.

So I agree we should return `.ip_addr = get_broadcast_address()`.

Comment (kbr-) [reply service/raft/raft_group0.cc:897]: Alternatively - is this `ip_addr` field even necessary?

Whoever sent the `peer_exchange` request - they obviously know the destination IP they sent it to. They can just reuse it. We're returning redundant information, no?

I guess it doesn't hurt to return `ip_addr` here, maybe it makes the code shorter on the sender side.

File: service/raft/raft_rpc.cc [hunk 17]
```diff
@@ -30,135 +32,99 @@ raft_rpc::raft_rpc(raft_state_machine& sm, netw::messaging_service& ms,
...
-    return ser::raft_rpc_verbs::send_raft_append_entries(&_messaging,
-        netw::msg_addr(_address_map.get_inet_address(id)), db::no_timeout, _group_id, _server_id, id, append_request);
-}
 
-void raft_rpc::send_append_entries_reply(raft::server_id id, const raft::append_reply& reply) {
-    (void)with_gate(_shutdown_gate, [this, id, &reply] {
-        return ser::raft_rpc_verbs::send_raft_append_entries_reply(&_messaging, netw::msg_addr(_address_map.get_inet_address(id)), timeout(), _group_id, _server_id, id, reply)
-            .handle_exception([id] (std::exception_ptr ex) {
+template <typename Verb, typename Msg> void
+raft_rpc::one_way_rpc(std::experimental::source_location loc, raft::server_id id,
```
Comment (kbr-) [inline service/raft/raft_rpc.cc:37]: Since you introduced the `sloc` alias, you could use it here.

File: service/raft/raft_rpc.cc [hunk 18]
```diff
@@ -30,135 +32,99 @@ raft_rpc::raft_rpc(raft_state_machine& sm, netw::messaging_service& ms,
...
-                    std::rethrow_exception(ex);
-                } catch (seastar::rpc::timeout_error&) {
-                } catch (seastar::rpc::closed_error&) {
-                } catch (...) {
-                    rlogger.error("Failed to send vote request {}: {}", id, ex);
-                }
-            });
+template <typename Verb, typename... Args>
+decltype(auto)
+raft_rpc::two_way_rpc(const std::experimental::source_location loc, raft::server_id id,
```
Comment (kbr-) [inline service/raft/raft_rpc.cc:61]: Since you introduced the `sloc` alias, you could use it here.

File: service/raft/raft_rpc.cc [hunk 19]
```diff
@@ -30,135 +32,99 @@ raft_rpc::raft_rpc(raft_state_machine& sm, netw::messaging_service& ms,
...
-                try {
-                    std::rethrow_exception(ex);
-                } catch (seastar::rpc::timeout_error&) {
-                } catch (seastar::rpc::closed_error&) {
-                } catch (...) {
-                    rlogger.error("Failed to send vote request {}: {}", id, ex);
-                }
-            });
+template <typename Verb, typename... Args>
+decltype(auto)
```
Comment (kbr-) [inline service/raft/raft_rpc.cc:60]: You can change it to `auto`, I checked that it compiles (unless you encounter runtime errors? That would be weird)

Obviously, you need to change it both in .cc and in .hh.

File: service/raft/raft_rpc.cc (hunk 18 above)
Comment (kbr-) [reply service/raft/raft_rpc.cc:61]: Also, you want to move the `loc` later, `const` prevents you from doing that.

File: service/raft/raft_address_map.hh [hunk 20]
```diff
@@ -205,6 +205,13 @@ class raft_address_map_t : public peering_sharded_service<raft_address_map_t<Clo
         });
     }
 
+    void replicate_add_or_update_entry(const raft::server_id& id, const gms::inet_address& ip_addr,
```
Comment (kbr-) [inline service/raft/raft_address_map.hh:208]: I thought we agreed that the caller should be responsible for setting mappings on the different shards...

**Review by kbr- — CHANGES_REQUESTED** [review]

File: service/raft/raft_address_map.hh (hunk 4 above)
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_address_map.hh:216]: Fixed.

File: service/raft/raft_address_map.hh [hunk 21]
```diff
@@ -263,101 +262,86 @@ public:
...
+    // Convert an expiring entry to a non-expiring one, or
+    // insert a new non-expiring entry if the entry is missing.
+    // Called on Raft configuration changes to mark the new
+    // member of Raft configuration as a non-temporary member
+    // of the address map. The configuration member may be
+    // lacking an IP address but it will be added later.
+    // Can only be called on shard 0.
+    // The expiring state is replicated to other shards.
+    void set_nonexpiring(raft::server_id id) {
+        auto [it, emplaced] = _map.try_emplace(std::move(id), timestamped_entry{{}});
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_address_map.hh]: Fixed.

File: service/raft/raft_address_map.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_address_map.hh]: Return value - gone, update_if_exists - turns out we indeed need it, added comments, left in plac.e

File: service/raft/raft_group0.cc [hunk 22]
```diff
@@ -86,6 +87,41 @@ namespace service {
 static logging::logger group0_log("raft_group0");
 static logging::logger upgrade_log("raft_group0_upgrade");
 
+// {{{ group0_rpc Maintain failure detector subscription whenever
+// group 0 configuration changes.
+
+class group0_rpc: public service::raft_rpc {
+    service::raft_group_registry& _raft_gr;
+public:
+    using raft_rpc::raft_rpc;
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_group0.cc:96]: Removed.

File: service/raft/raft_group0.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_group0.cc:383]: Fixed, pass 'false' as update_if_exists.

File: service/raft/raft_group_registry.cc (hunk 1 above)
Comment (kostja) [reply service/raft/raft_group_registry.cc:167]: Removed.

File: service/raft/raft_group0.cc (hunk 16 above)
Comment (kostja) [reply service/raft/raft_group0.cc:897]: It makes life easier for the receiver. I'll use get_broadcast_addr()

Comment (kostja) [reply service/raft/raft_group0.cc:897]: So the history of this is that it would initially return the leader id and address. This was removed when forwarding was added, assuming our own address is just as good as hte leader's. this adds an extra hop. But having the ip and addr explicitly in response also adds hope that the hop can be removed in the future.

File: service/raft/raft_rpc.cc (hunk 17 above)
Comment (kostja) [reply service/raft/raft_rpc.cc:37]: Done.

File: service/raft/raft_rpc.cc (hunk 18 above)
Comment (kostja) [reply service/raft/raft_rpc.cc:61]: Done.

File: service/raft/raft_rpc.cc [hunk 23]
```diff
@@ -30,135 +32,99 @@ raft_rpc::raft_rpc(raft_state_machine& sm, netw::messaging_service& ms,
...
+raft_rpc::two_way_rpc(const std::experimental::source_location loc, raft::server_id id,
+        Verb&& verb, Args&&... args) {
+    auto ip_addr = _address_map.find(id);
+    using Fut = decltype(verb(&_messaging, netw::msg_addr(*ip_addr), db::no_timeout, _group_id, _server_id, id, std::forward<Args>(args)...));
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_rpc.cc:64]: I don't get it how it's shorter. I'll have to list all type names explicitly. Here I list onlythe variable names. Could you please provide an example?

File: service/raft/raft_rpc.cc [hunk 24]
```diff
@@ -30,135 +32,99 @@ raft_rpc::raft_rpc(raft_state_machine& sm, netw::messaging_service& ms,
...
+    using Ret = typename Fut::value_type;
+    if (!ip_addr) {
+        const auto msg = format("Failed to send {} {}: ip address not found", loc.function_name(), id);
+        return make_exception_future<Ret>(raft::transport_error(msg));
+    }
+    return verb(&_messaging, netw::msg_addr(*ip_addr), db::no_timeout, _group_id, _server_id, id, std::forward<Args>(args)...)
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_rpc.cc:70]: I believe because you can't really abort a two-way RPC in our RPC impl today. I.e. you can pretend it's aborted, but there has to be a future somewhere which reads the result. So it doesn't make sense. @gleb-cloudius I remember we discussed it with you some time ago.

File: service/raft/raft_rpc.cc [hunk 25]
```diff
@@ -30,135 +32,99 @@ raft_rpc::raft_rpc(raft_state_machine& sm, netw::messaging_service& ms,
...
+        Verb&& verb, Msg&& msg) {
+    (void)with_gate(_shutdown_gate, [this, loc = std::move(loc), id, &verb, &msg] {
+        auto ip_addr = _address_map.find(id);
+        if (!ip_addr) {
+            rlogger.debug("{}:{}: {} dropping outgoing message to {} - IP address not found",
+                loc.file_name(), loc.line(), loc.function_name(), id);
+            return make_ready_future<>();
+        }
+        return verb(&_messaging, netw::msg_addr(*ip_addr), timeout(), _group_id, _server_id, id, std::forward<Msg>(msg))
+            .handle_exception([loc = std::move(loc), id] (std::exception_ptr ex) {
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_rpc.cc:47]: It's a bunch of sstrings.

File: service/raft/raft_address_map.hh (hunk 20 above)
Comment (kostja) [reply service/raft/raft_address_map.hh:208]: Discussed and resolved we have to keep it.

File: service/raft/raft_group0.cc (hunk 16 above)
Comment (kostja) [reply service/raft/raft_group0.cc]: Used _gossiper.get_broadcast_address()

File: service/raft/raft_rpc.cc (hunk 19 above)
Comment (kostja) [reply service/raft/raft_rpc.cc:60]: @kbr- next time plesae if you make a comment please provide a rationale. I changed now to your liking, but afaiu decltype() was added to the standard specifically for return value declarations, so it was used appropriately. for the sake of making progress, I am accepting your remark and moving on.

File: service/raft/raft_rpc.cc (hunk 23 above)
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_rpc.cc:64]: I assume it doesn't make much a difference, and the current approach is shorter. Closing with no action.

File: service/raft/raft_rpc.hh [hunk 26]
```diff
@@ -33,6 +33,12 @@ public:
     explicit raft_rpc(raft_state_machine& sm, netw::messaging_service& ms,
             raft_address_map& address_map, raft::group_id gid, raft::server_id srv_id);
 
+    template <typename Verb, typename Msg> void
+    one_way_rpc(std::experimental::source_location loc, raft::server_id id, Verb&& verb, Msg&& msg);
+
+    template <typename Verb, typename... Args> decltype(auto)
+    two_way_rpc(std::experimental::source_location loc, raft::server_id id, Verb&& verb, Args&&... args);
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_rpc.hh:41]: Done.

File: db/system_keyspace.cc [hunk 27]
```diff
@@ -1543,6 +1543,20 @@ future<std::unordered_map<gms::inet_address, locator::host_id>> system_keyspace:
     });
 }
 
+future<std::unordered_map<gms::inet_address, utils::UUID>> system_keyspace::load_raft_server_ids() {
+    sstring req = format("SELECT peer, raft_server_id FROM system.{}", PEERS);
+    shared_ptr<cql3::untyped_result_set> cql_result = co_await qctx->execute_cql(req);
```
[the remark this answers is not in the document]
Comment (kostja) [reply db/system_keyspace.cc:1548]: Done.

File: service/raft/raft_rpc.hh [hunk 28]
```diff
@@ -11,33 +11,27 @@
...
-    noncopyable_function<void(raft::server_id, bool added)> _on_server_update;
+    raft_address_map& _address_map;
     seastar::gate _shutdown_gate;
-
-    raft_ticker_type::time_point timeout() {
-        return raft_ticker_type::clock::now() + raft_tick_interval * (raft::ELECTION_TIMEOUT.count() / 2);
-    }
-
 public:
     explicit raft_rpc(raft_state_machine& sm, netw::messaging_service& ms,
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_rpc.hh:33]: Done.

File: service/raft/raft_address_map.hh [hunk 29]
```diff
@@ -263,101 +262,86 @@ public:
...
-            replicate_set_expiring_flag(it->first);
+            return;
         }
-        return entry._addr;
+        add_expiring_entry(it->first, entry);
+        replicate_set_expiring(it->first);
     }
+    // Inserts a new expiring mapping or updates the existing one
+    // with the new IP address. Never clears the expiring property
+    // from an existing entry.
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_address_map.hh]: Done.

File: service/raft/raft_group0.cc [hunk 30]
```diff
@@ -86,6 +87,41 @@ namespace service {
...
+    virtual void on_configuration_change(raft::server_address_set add, raft::server_address_set del) override;
+    explicit group0_rpc(raft_group_registry& raft_gr,
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_group0.cc]: I could take time to finish this https://github.com/scylladb/scylladb/commit/6cac36418a2657bcb716ccfeaab69392dbde9adf in a follow up commit if you approve, for now I'll keep the current code.

File: service/raft/raft_rpc.cc (hunk 25 above)
[the remark this answers is not in the document]
Comment (kbr-) [reply service/raft/raft_rpc.cc:47]: I think it's not specified what exactly they are. A string could be `std::string` (expensive copies) or it could be `std::string_view` (cheap copies).

(Although I doubt `source_location` performs an allocation, it probably is just a pointer to a string living somewhere in .rodata or similar)

But let's stay on the safe side and ensure they are always moved.

**Review by kbr- — CHANGES_REQUESTED** [review]

File: service/raft/raft_address_map.hh [hunk 31]
```diff
@@ -263,101 +284,63 @@ public:
...
+        handle_add_or_update_entry(id, addr, false);
     }
-    raft::server_address get_server_address(raft::server_id id) const {
-        return raft::server_address{id, ser::serialize_to_buffer<bytes>(get_inet_address(id))};
+    // Insert or update entry with a new IP address on all shards.
+    // Used when we get a gossip notification about a node IP
+    // address. Overrides the current IP address if present.
+    // If no entry is present, creates an expiring entry - there
+    // must be a separate Raft configuration change event (@sa
+    // set_nonexpiring()) to mark the entry as expiring.
```
[the remark this answers is not in the document]
Comment (kostja) [reply service/raft/raft_address_map.hh:334]: Done.

File: test/raft/raft_address_map_test.cc [hunk 32]
```diff
@@ -52,79 +52,60 @@ SEASTAR_THREAD_TEST_CASE(test_raft_address_map_operations) {
...
-        m.set(id1, addr1, true);
         {
             const auto found = m.find(id1);
             BOOST_CHECK(!!found && *found == addr1);
         }
-        {
-            const auto found = m.set_expiring_flag(id1);
-            BOOST_CHECK(!!found && *found == addr1);
-        }
+        m.set_expiring(id1);
```
[the remark this answers is not in the document]
Comment (kostja) [reply test/raft/raft_address_map_test.cc:68]: OK.

File: test/raft/raft_address_map_test.cc (hunk 10 above)
Comment (kostja) [reply test/raft/raft_address_map_test.cc:94]: Removed.

File: test/raft/raft_address_map_test.cc (hunk 11 above)
Comment (kostja) [reply test/raft/raft_address_map_test.cc:128]: Re test failing anyway if it throws - I think the entire test would fail, not just one case, which is undesirable. Will add a test that update works.

Comment (kostja) [reply test/raft/raft_address_map_test.cc:128]: Done.

File: test/raft/raft_address_map_test.cc (hunk 12 above)
Comment (kostja) [reply test/raft/raft_address_map_test.cc:142]: Done.

**Review by kbr- — APPROVED** [review]
LGTM

@tgrabiec, please review as well.

**Comment by kbr-** [comment]
That last change probably deserves a test, but we can do it in a follow-up.
https://github.com/scylladb/scylladb/compare/765dbffbaa3edfe8482376ac43bbd3cd71f03050..fbe7886cc0cf252b2e0716d637fd6b0fcc3a571f

**Comment by kbr-** [comment]
I'm merging to unblock some dependencies.

Known follow-up items:
- more unit tests for raft address map
- IP address change testing in test.py topology tests
- optimize so we don't load every node's RAFT ID from local table on bootstrap separately in `gossiper::add_saved_endpoint`. We should load the table into memory using a single scan (like we do with host ids and tokens) and use the in-memory struct in `gossiper::add_saved_endpoint`
- @kostja maybe I forgot something, please add

Any further review can also be addressed in a follow-up.
